### PR TITLE
add QL specification section on module instantiations

### PR DESCRIPTION
**Pull request by ginsbach**
- fixes https://github.com/github/codeql-core/issues/2452

### CODE REVIEW

**Review by alexet — DISMISSED** [review]
LGTM but I think there are still some missing parts

For all the bits in the evaluation sections that kind-of happen "after substitution" we probably want something to just state that we handle predicates by using the normal algorithm but replacing all `instantiation-relative` predicates and types with the actual version. You probably want to say this should happen after we resolve calls and type references but I think we can get away with being a bit vague about precisely what that means. 

The section on layer evaluation may want a note that it applies to "fully instantiated" types and predicates.

Otherwise the missing parts are really covered by the general section on components.

**Comment by ginsbach** [comment]
> The section on layer evaluation may want a note that it applies to "fully instantiated" types and predicates.

I have added this in the fifth commit.

> For all the bits in the evaluation sections that kind-of happen "after substitution" we probably want something to just state that we handle predicates by using the normal algorithm but replacing all `instantiation-relative` predicates and types with the actual version. You probably want to say this should happen after we resolve calls and type references but I think we can get away with being a bit vague about precisely what that means.

Can you elaborate a bit which section you mean and what changes you'd like to see?

File: docs/codeql/ql-language-reference/ql-language-specification.rst [hunk 1]
```diff
@@ -2040,7 +2042,7 @@ Stratification
 
 A QL program can be *stratified* to a sequence of *layers*. A layer is a set of predicates and types.
 
-A valid stratification must include each predicate and type in the QL program. It must not include any other predicates or types.
+A valid stratification must include each predicate and type in the QL program that is *fully instantiated*. It must not include any other predicates or types.
```
Comment (alexet) [inline docs/codeql/ql-language-reference/ql-language-specification.rst:2045]: I'm not too bothered but we also check the uninstantiated cases as well so we should include them. (on the other hand we would need to state that we ignore dispatch for parameters).

Comment (alexet) [reply docs/codeql/ql-language-reference/ql-language-specification.rst:2045]: Although we would also need to state evaluation only happens on *fully instantiated* predicates

**Review by alexet — APPROVED** [review]
Changes LGTM

The point I was trying to make is that the sections on `Stratification` and `Layer evaluation` only really apply to declared predicates. We don't need to go into great detail but a line saying it works by substituting after resolving is fine. Alternatively we could probably get away with by substituting after resolving types and predicates.

File: docs/codeql/ql-language-reference/ql-language-specification.rst (hunk 1 above)
Comment (ginsbach) [reply docs/codeql/ql-language-reference/ql-language-specification.rst:2045]: We report errors on uninstantiated cases as well, that is true. However, given an error-free QL program, the difference between adding in the uninstantiated cases in the stratification (and filtering out per layer afterwards) or not should not be observable. Therefore, I think we can get away with this here.

**Comment by ginsbach** [comment]
> Changes LGTM
> 
> The point I was trying to make is that the sections on `Stratification` and `Layer evaluation` only really apply to declared predicates. We don't need to go into great detail but a line saying it works by substituting after resolving is fine. Alternatively we could probably get away with by substituting after resolving types and predicates.

I will consider a follow-up PR to polish the evaluation section.

**Pull request merged**